In [88]:
#import working library
import pandas as pd


In [89]:
#Import our first data source for observation, Investigation and cleaning
df = pd.read_csv(r"C:\Users\HEY\Documents\Madebyenigma\Projects\Data Analytics Portfolio\Google Capstone\Bike Sharing Exercise\2. Data\RAW\Divvy_Trips_2019_Q1.csv")

In [90]:
#
df.head()

,trip_id,start_time,end_time,bikeid,tripduration,from_station_id,from_station_name,to_station_id,to_station_name,usertype,gender,birthyear
0,21742443,2019-01-01 00:04:37,2019-01-01 00:11:07,2167,390.0,199,Wabash Ave & Grand Ave,84,Milwaukee Ave & Grand Ave,Subscriber,Male,1989.0
1,21742444,2019-01-01 00:08:13,2019-01-01 00:15:34,4386,441.0,44,State St & Randolph St,624,Dearborn St & Van Buren St (*),Subscriber,Female,1990.0
2,21742445,2019-01-01 00:13:23,2019-01-01 00:27:12,1524,829.0,15,Racine Ave & 18th St,644,Western Ave & Fillmore St (*),Subscriber,Female,1994.0
3,21742446,2019-01-01 00:13:45,2019-01-01 00:43:28,252,"1,783.0",123,California Ave & Milwaukee Ave,176,Clark St & Elm St,Subscriber,Male,1993.0
4,21742447,2019-01-01 00:14:52,2019-01-01 00:20:56,1170,364.0,173,Mies van der Rohe Way & Chicago Ave,35,Streeter Dr & Grand Ave,Subscriber,Male,1994.0


In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365069 entries, 0 to 365068
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   trip_id            365069 non-null  int64  
 1   start_time         365069 non-null  object 
 2   end_time           365069 non-null  object 
 3   bikeid             365069 non-null  int64  
 4   tripduration       365069 non-null  object 
 5   from_station_id    365069 non-null  int64  
 6   from_station_name  365069 non-null  object 
 7   to_station_id      365069 non-null  int64  
 8   to_station_name    365069 non-null  object 
 9   usertype           365069 non-null  object 
 10  gender             345358 non-null  object 
 11  birthyear          347046 non-null  float64
dtypes: float64(1), int64(4), object(7)
memory usage: 33.4+ MB


In [92]:
print(df.isnull().sum())

trip_id                  0
start_time               0
end_time                 0
bikeid                   0
tripduration             0
from_station_id          0
from_station_name        0
to_station_id            0
to_station_name          0
usertype                 0
gender               19711
birthyear            18023
dtype: int64


In [93]:
df['tripduration'].head(10)

0      390.0
1      441.0
2      829.0
3    1,783.0
4      364.0
5      216.0
6      177.0
7      100.0
8    1,727.0
9      336.0
Name: tripduration, dtype: object

In [94]:
df['tripduration'].head(30)

0       390.0
1       441.0
2       829.0
3     1,783.0
4       364.0
5       216.0
6       177.0
7       100.0
8     1,727.0
9       336.0
10      886.0
11      653.0
12      601.0
13      562.0
14      906.0
15      892.0
16      407.0
17    1,356.0
18      102.0
19    2,333.0
20      960.0
21    2,301.0
22    1,168.0
23      474.0
24    1,036.0
25      230.0
26      747.0
27      711.0
28      553.0
29      521.0
Name: tripduration, dtype: object

In [95]:
# Confrim data type for input without comma
type(df['tripduration'].iloc[0])

str

In [96]:
# Comfirm data type for input with comma
type(df['tripduration'].iloc[3])

str

In [97]:
# Confirm how many columns had input with commas in them
df['tripduration'].str.contains(',', na=False).sum()

np.int64(72042)

In [98]:
# check for the data type of the start time column
type(df['start_time'].iloc[0])

str

In [99]:
# check for the data type of the end time column
type(df['end_time'].iloc[0])

str

# Data Quality Assessment 

## Dataset Overview

The Divvy Trips 2019 Q1 dataset contains 365,069 records and 12 columns.

Initial exploration was conducted using:
- df.shape
- df.info()
- df.isnull().sum()

The objective was to identify missing values, data type inconsistencies, and other potential data quality issues before cleaning.

## Findings

### Missing Values

The following columns contain missing values:

| Column | Missing Values |
|----------|----------|
| gender | 19,711 |
| birthyear | 18,023 |

All other columns contain no missing values.

### Data Type Issues

The following columns were identified as having potentially incorrect data types:

| Column | Current Type | Expected Type |
|----------|----------|----------|
| start_time | object (string) | datetime |
| end_time | object (string) | datetime |
| tripduration | object (string) | numeric |

### Trip Duration Investigation

Further investigation revealed that the tripduration column contains values formatted with comma separators (e.g., 1,783.0).

Approximately 72,042 records contain comma-separated values, which likely caused Pandas to interpret the column as a string rather than a numeric datatype.

### Schema Consistency Observation

The 2019 and 2020 datasets use different schemas.

The 2019 dataset contains demographic fields such as:
- gender
- birthyear

These fields are not present in the 2020 dataset and may require special consideration during data integration.

In [100]:
#Before deciding on cleaning, check the format of the date and time
df[['start_time', 'end_time']].head()

,start_time,end_time
0,2019-01-01 00:04:37,2019-01-01 00:11:07
1,2019-01-01 00:08:13,2019-01-01 00:15:34
2,2019-01-01 00:13:23,2019-01-01 00:27:12
3,2019-01-01 00:13:45,2019-01-01 00:43:28
4,2019-01-01 00:14:52,2019-01-01 00:20:56


## Planned Cleaning Steps

1. Convert start_time to datetime format.
2. Convert end_time to datetime format.
3. Remove comma formatting from tripduration.
4. Convert tripduration to a numeric datatype.
5. Evaluate whether gender and birthyear should be retained.
6. Standardize column names for compatibility with 2020 data.

In [101]:
#Import our first data source for observation, Investigation and cleaning
df_2020 = pd.read_csv(r"C:\Users\HEY\Documents\Madebyenigma\Projects\Data Analytics Portfolio\Google Capstone\Bike Sharing Exercise\2. Data\RAW\Divvy_Trips_2020_Q1.csv")

In [102]:
df_2020.head()

,ride_id,rideable_type,started_at,ended_at,start_station_name,start_station_id,end_station_name,end_station_id,start_lat,start_lng,end_lat,end_lng,member_casual
0,EACB19130B0CDA4A,docked_bike,2020-01-21 20:06:59,2020-01-21 20:14:30,Western Ave & Leland Ave,239,Clark St & Leland Ave,326.0,41.9665,-87.6884,41.9671,-87.6674,member
1,8FED874C809DC021,docked_bike,2020-01-30 14:22:39,2020-01-30 14:26:22,Clark St & Montrose Ave,234,Southport Ave & Irving Park Rd,318.0,41.9616,-87.6660,41.9542,-87.6644,member
2,789F3C21E472CA96,docked_bike,2020-01-09 19:29:26,2020-01-09 19:32:17,Broadway & Belmont Ave,296,Wilton Ave & Belmont Ave,117.0,41.9401,-87.6455,41.9402,-87.6530,member
3,C9A388DAC6ABF313,docked_bike,2020-01-06 16:17:07,2020-01-06 16:25:56,Clark St & Randolph St,51,Fairbanks Ct & Grand Ave,24.0,41.8846,-87.6319,41.8918,-87.6206,member
4,943BC3CBECCFD662,docked_bike,2020-01-30 08:37:16,2020-01-30 08:42:48,Clinton St & Lake St,66,Wells St & Hubbard St,212.0,41.8856,-87.6418,41.8899,-87.6343,member


In [103]:
df_2020.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426887 entries, 0 to 426886
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             426887 non-null  object 
 1   rideable_type       426887 non-null  object 
 2   started_at          426887 non-null  object 
 3   ended_at            426887 non-null  object 
 4   start_station_name  426887 non-null  object 
 5   start_station_id    426887 non-null  int64  
 6   end_station_name    426886 non-null  object 
 7   end_station_id      426886 non-null  float64
 8   start_lat           426887 non-null  float64
 9   start_lng           426887 non-null  float64
 10  end_lat             426886 non-null  float64
 11  end_lng             426886 non-null  float64
 12  member_casual       426887 non-null  object 
dtypes: float64(5), int64(1), object(7)
memory usage: 42.3+ MB


In [104]:
df_2020.shape

(426887, 13)

In [105]:
df_2020.isnull().sum()

ride_id               0
rideable_type         0
started_at            0
ended_at              0
start_station_name    0
start_station_id      0
end_station_name      1
end_station_id        1
start_lat             0
start_lng             0
end_lat               1
end_lng               1
member_casual         0
dtype: int64

# Key Findings

After assessing both the 2019 and 2020 datasets, the following issues were identified:

1. Datetime fields are stored as strings in both datasets.
2. The 2019 tripduration field is stored as a string and contains comma-formatted values.
3. The 2019 and 2020 datasets use different schemas and naming conventions.
4. The 2019 dataset contains demographic fields not available in 2020.
5. Missing values are minimal and largely concentrated in demographic fields.

In [106]:
#Create a copy of the dataset 
df_clean = df.copy()

In [107]:
df_clean['start_time'] = pd.to_datetime(df_clean['start_time'])

df_clean['end_time'] = pd.to_datetime(df_clean['end_time'])

In [108]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365069 entries, 0 to 365068
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   trip_id            365069 non-null  int64         
 1   start_time         365069 non-null  datetime64[ns]
 2   end_time           365069 non-null  datetime64[ns]
 3   bikeid             365069 non-null  int64         
 4   tripduration       365069 non-null  object        
 5   from_station_id    365069 non-null  int64         
 6   from_station_name  365069 non-null  object        
 7   to_station_id      365069 non-null  int64         
 8   to_station_name    365069 non-null  object        
 9   usertype           365069 non-null  object        
 10  gender             345358 non-null  object        
 11  birthyear          347046 non-null  float64       
dtypes: datetime64[ns](2), float64(1), int64(4), object(5)
memory usage: 33.4+ MB


# Cleaning Step 1: Datetime Conversion

The start_time and end_time columns were originally stored as strings.

Both columns were successfully converted to datetime format using pd.to_datetime().

Verification through df_clean.info() confirmed that both fields now use the datetime64[ns] datatype.

This transformation enables future time-based analysis, including monthly, weekly, and hourly usage patterns.

In [109]:
#Check the data one more time in this version just to be sure
df_clean['tripduration'].head(10)

0      390.0
1      441.0
2      829.0
3    1,783.0
4      364.0
5      216.0
6      177.0
7      100.0
8    1,727.0
9      336.0
Name: tripduration, dtype: object

In [110]:
# Confirm data type also is still string
type(df_clean['tripduration'].iloc[0])

str

In [111]:
#First we remove the commas
df_clean['tripduration'] = df_clean['tripduration'].str.replace(',', '', regex=False)

In [112]:
#Verify that all commas are gone
df_clean['tripduration'].str.contains(',', na=False).sum()

np.int64(0)

In [113]:
#Convert data type to a numeric data type
df_clean['tripduration'] = pd.to_numeric(df_clean['tripduration'])

In [114]:
#Confirm data type change
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365069 entries, 0 to 365068
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   trip_id            365069 non-null  int64         
 1   start_time         365069 non-null  datetime64[ns]
 2   end_time           365069 non-null  datetime64[ns]
 3   bikeid             365069 non-null  int64         
 4   tripduration       365069 non-null  float64       
 5   from_station_id    365069 non-null  int64         
 6   from_station_name  365069 non-null  object        
 7   to_station_id      365069 non-null  int64         
 8   to_station_name    365069 non-null  object        
 9   usertype           365069 non-null  object        
 10  gender             345358 non-null  object        
 11  birthyear          347046 non-null  float64       
dtypes: datetime64[ns](2), float64(2), int64(4), object(4)
memory usage: 33.4+ MB


## Cleaning Progress

### Datetime Conversion

The start_time and end_time columns were originally stored as strings.

After verifying that the values followed a consistent datetime format, both columns were converted using pd.to_datetime().

This allows for future time-based analysis such as identifying ride patterns by month, day of week, and hour.

### Trip Duration Conversion

The tripduration column was initially stored as a string rather than a numeric datatype.

Further investigation revealed that approximately 72,042 records contained comma separators (e.g., 1,783.0), which prevented Pandas from interpreting the column as numeric.

The commas were removed and the column was converted to a numeric datatype using pd.to_numeric().

Verification confirmed that the column was successfully converted to float64.

In [115]:
# Check to ensure there are no duplicated data
df_clean.duplicated().sum()

np.int64(0)

In [116]:
# Import 2019_Q2 so we can Verify the schema
df_q2 = pd.read_csv(
    r"C:\Users\HEY\Documents\Madebyenigma\Projects\Data Analytics Portfolio\Google Capstone\Bike Sharing Exercise\2. Data\RAW\Divvy_Trips_2019_Q2.csv"
)

ERROR! Session/line number was not unique in database. History logging moved to new session 2


In [ ]:
df_q2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1108163 entries, 0 to 1108162
Data columns (total 12 columns):
 #   Column                                            Non-Null Count    Dtype  
---  ------                                            --------------    -----  
 0   01 - Rental Details Rental ID                     1108163 non-null  int64  
 1   01 - Rental Details Local Start Time              1108163 non-null  object 
 2   01 - Rental Details Local End Time                1108163 non-null  object 
 3   01 - Rental Details Bike ID                       1108163 non-null  int64  
 4   01 - Rental Details Duration In Seconds Uncapped  1108163 non-null  object 
 5   03 - Rental Start Station ID                      1108163 non-null  int64  
 6   03 - Rental Start Station Name                    1108163 non-null  object 
 7   02 - Rental End Station ID                        1108163 non-null  int64  
 8   02 - Rental End Station Name                      1108163 non-null  obje

In [ ]:
df_q2.head()

,01 - Rental Details Rental ID,01 - Rental Details Local Start Time,01 - Rental Details Local End Time,01 - Rental Details Bike ID,01 - Rental Details Duration In Seconds Uncapped,03 - Rental Start Station ID,03 - Rental Start Station Name,02 - Rental End Station ID,02 - Rental End Station Name,User Type,Member Gender,05 - Member Details Member Birthday Year
0,22178529,2019-04-01 00:02:22,2019-04-01 00:09:48,6251,446.0,81,Daley Center Plaza,56,Desplaines St & Kinzie St,Subscriber,Male,1975.0
1,22178530,2019-04-01 00:03:02,2019-04-01 00:20:30,6226,"1,048.0",317,Wood St & Taylor St,59,Wabash Ave & Roosevelt Rd,Subscriber,Female,1984.0
2,22178531,2019-04-01 00:11:07,2019-04-01 00:15:19,5649,252.0,283,LaSalle St & Jackson Blvd,174,Canal St & Madison St,Subscriber,Male,1990.0
3,22178532,2019-04-01 00:13:01,2019-04-01 00:18:58,4151,357.0,26,McClurg Ct & Illinois St,133,Kingsbury St & Kinzie St,Subscriber,Male,1993.0
4,22178533,2019-04-01 00:19:26,2019-04-01 00:36:13,3270,"1,007.0",202,Halsted St & 18th St,129,Blue Island Ave & 18th St,Subscriber,Male,1992.0


In [ ]:
# Import 2019_Q3 so we can Verify the schema
df_q3 = pd.read_csv(
    r"C:\Users\HEY\Documents\Madebyenigma\Projects\Data Analytics Portfolio\Google Capstone\Bike Sharing Exercise\2. Data\RAW\Divvy_Trips_2019_Q3.csv"
)

In [ ]:
df_q3.head()

,trip_id,start_time,end_time,bikeid,tripduration,from_station_id,from_station_name,to_station_id,to_station_name,usertype,gender,birthyear
0,23479388,2019-07-01 00:00:27,2019-07-01 00:20:41,3591,"1,214.0",117,Wilton Ave & Belmont Ave,497,Kimball Ave & Belmont Ave,Subscriber,Male,1992.0
1,23479389,2019-07-01 00:01:16,2019-07-01 00:18:44,5353,"1,048.0",381,Western Ave & Monroe St,203,Western Ave & 21st St,Customer,NaN,NaN
2,23479390,2019-07-01 00:01:48,2019-07-01 00:27:42,6180,"1,554.0",313,Lakeview Ave & Fullerton Pkwy,144,Larrabee St & Webster Ave,Customer,NaN,NaN
3,23479391,2019-07-01 00:02:07,2019-07-01 00:27:10,5540,"1,503.0",313,Lakeview Ave & Fullerton Pkwy,144,Larrabee St & Webster Ave,Customer,NaN,NaN
4,23479392,2019-07-01 00:02:13,2019-07-01 00:22:26,6014,"1,213.0",168,Michigan Ave & 14th St,62,McCormick Place,Customer,NaN,NaN


In [ ]:
df_q3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1640718 entries, 0 to 1640717
Data columns (total 12 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   trip_id            1640718 non-null  int64  
 1   start_time         1640718 non-null  object 
 2   end_time           1640718 non-null  object 
 3   bikeid             1640718 non-null  int64  
 4   tripduration       1640718 non-null  object 
 5   from_station_id    1640718 non-null  int64  
 6   from_station_name  1640718 non-null  object 
 7   to_station_id      1640718 non-null  int64  
 8   to_station_name    1640718 non-null  object 
 9   usertype           1640718 non-null  object 
 10  gender             1353368 non-null  object 
 11  birthyear          1362624 non-null  float64
dtypes: float64(1), int64(4), object(7)
memory usage: 150.2+ MB


In [ ]:
# Import 2019_Q4 so we can Verify the schema
df_q4 = pd.read_csv(
    r"C:\Users\HEY\Documents\Madebyenigma\Projects\Data Analytics Portfolio\Google Capstone\Bike Sharing Exercise\2. Data\RAW\Divvy_Trips_2019_Q4.csv"
)

In [ ]:
df_q4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 704054 entries, 0 to 704053
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   trip_id            704054 non-null  int64  
 1   start_time         704054 non-null  object 
 2   end_time           704054 non-null  object 
 3   bikeid             704054 non-null  int64  
 4   tripduration       704054 non-null  object 
 5   from_station_id    704054 non-null  int64  
 6   from_station_name  704054 non-null  object 
 7   to_station_id      704054 non-null  int64  
 8   to_station_name    704054 non-null  object 
 9   usertype           704054 non-null  object 
 10  gender             637463 non-null  object 
 11  birthyear          642373 non-null  float64
dtypes: float64(1), int64(4), object(7)
memory usage: 64.5+ MB


## Q2 Column Mapping

01 - Rental Details Rental ID
- trip_id

01 - Rental Details Local Start Time
- start_time

01 - Rental Details Local End Time
- end_time

01 - Rental Details Bike ID
- bikeid

01 - Rental Details Duration In Seconds Uncapped
- tripduration

03 - Rental Start Station ID
- from_station_id

03 - Rental Start Station Name
-  from_station_name

02 - Rental End Station ID
- to_station_id

02 - Rental End Station Name
- to_station_name

User Type
- usertype

Member Gender
- gender

05 - Member Details Member Birthday Year
→ birthyear

In [ ]:
# Create a copy of q2 to rename 
df_q2_clean = df_q2.copy()

In [ ]:
#Rename the column headers to match the other datasets
df_q2_clean = df_q2_clean.rename(columns={
    '01 - Rental Details Rental ID': 'trip_id',
    '01 - Rental Details Local Start Time': 'start_time',
    '01 - Rental Details Local End Time': 'end_time',
    '01 - Rental Details Bike ID': 'bikeid',
    '01 - Rental Details Duration In Seconds Uncapped': 'tripduration',
    '03 - Rental Start Station ID': 'from_station_id',
    '03 - Rental Start Station Name': 'from_station_name',
    '02 - Rental End Station ID': 'to_station_id',
    '02 - Rental End Station Name': 'to_station_name',
    'User Type': 'usertype',
    'Member Gender': 'gender',
    '05 - Member Details Member Birthday Year': 'birthyear'
})

In [ ]:
#Verify change
df_q2_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1108163 entries, 0 to 1108162
Data columns (total 12 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   trip_id            1108163 non-null  int64  
 1   start_time         1108163 non-null  object 
 2   end_time           1108163 non-null  object 
 3   bikeid             1108163 non-null  int64  
 4   tripduration       1108163 non-null  object 
 5   from_station_id    1108163 non-null  int64  
 6   from_station_name  1108163 non-null  object 
 7   to_station_id      1108163 non-null  int64  
 8   to_station_name    1108163 non-null  object 
 9   usertype           1108163 non-null  object 
 10  gender             922609 non-null   object 
 11  birthyear          927210 non-null   float64
dtypes: float64(1), int64(4), object(7)
memory usage: 101.5+ MB


In [ ]:
# create a copy for q3 data set
df_q3_clean = df_q3.copy()

In [ ]:
# check that the copy has all fields intact with the data types still as objects
df_q3_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1640718 entries, 0 to 1640717
Data columns (total 12 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   trip_id            1640718 non-null  int64  
 1   start_time         1640718 non-null  object 
 2   end_time           1640718 non-null  object 
 3   bikeid             1640718 non-null  int64  
 4   tripduration       1640718 non-null  object 
 5   from_station_id    1640718 non-null  int64  
 6   from_station_name  1640718 non-null  object 
 7   to_station_id      1640718 non-null  int64  
 8   to_station_name    1640718 non-null  object 
 9   usertype           1640718 non-null  object 
 10  gender             1353368 non-null  object 
 11  birthyear          1362624 non-null  float64
dtypes: float64(1), int64(4), object(7)
memory usage: 150.2+ MB


In [ ]:
# create a copy for q4 data set
df_q4_clean = df_q4.copy()

In [ ]:
# check that the copy has all fields intact with the data types still as objects
df_q4_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 704054 entries, 0 to 704053
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   trip_id            704054 non-null  int64  
 1   start_time         704054 non-null  object 
 2   end_time           704054 non-null  object 
 3   bikeid             704054 non-null  int64  
 4   tripduration       704054 non-null  object 
 5   from_station_id    704054 non-null  int64  
 6   from_station_name  704054 non-null  object 
 7   to_station_id      704054 non-null  int64  
 8   to_station_name    704054 non-null  object 
 9   usertype           704054 non-null  object 
 10  gender             637463 non-null  object 
 11  birthyear          642373 non-null  float64
dtypes: float64(1), int64(4), object(7)
memory usage: 64.5+ MB


In [ ]:
# check if q4 had the same issue as q1 with tripduration having commas
df_q4_clean.head()

,trip_id,start_time,end_time,bikeid,tripduration,from_station_id,from_station_name,to_station_id,to_station_name,usertype,gender,birthyear
0,25223640,2019-10-01 00:01:39,2019-10-01 00:17:20,2215,940.0,20,Sheffield Ave & Kingsbury St,309,Leavitt St & Armitage Ave,Subscriber,Male,1987.0
1,25223641,2019-10-01 00:02:16,2019-10-01 00:06:34,6328,258.0,19,Throop (Loomis) St & Taylor St,241,Morgan St & Polk St,Subscriber,Male,1998.0
2,25223642,2019-10-01 00:04:32,2019-10-01 00:18:43,3003,850.0,84,Milwaukee Ave & Grand Ave,199,Wabash Ave & Grand Ave,Subscriber,Female,1991.0
3,25223643,2019-10-01 00:04:32,2019-10-01 00:43:43,3275,"2,350.0",313,Lakeview Ave & Fullerton Pkwy,290,Kedzie Ave & Palmer Ct,Subscriber,Male,1990.0
4,25223644,2019-10-01 00:04:34,2019-10-01 00:35:42,5294,"1,867.0",210,Ashland Ave & Division St,382,Western Ave & Congress Pkwy,Subscriber,Male,1987.0


In [ ]:
# check if q3 had the same issue as q1 with tripduration having commas
df_q3_clean.head()

,trip_id,start_time,end_time,bikeid,tripduration,from_station_id,from_station_name,to_station_id,to_station_name,usertype,gender,birthyear
0,23479388,2019-07-01 00:00:27,2019-07-01 00:20:41,3591,"1,214.0",117,Wilton Ave & Belmont Ave,497,Kimball Ave & Belmont Ave,Subscriber,Male,1992.0
1,23479389,2019-07-01 00:01:16,2019-07-01 00:18:44,5353,"1,048.0",381,Western Ave & Monroe St,203,Western Ave & 21st St,Customer,NaN,NaN
2,23479390,2019-07-01 00:01:48,2019-07-01 00:27:42,6180,"1,554.0",313,Lakeview Ave & Fullerton Pkwy,144,Larrabee St & Webster Ave,Customer,NaN,NaN
3,23479391,2019-07-01 00:02:07,2019-07-01 00:27:10,5540,"1,503.0",313,Lakeview Ave & Fullerton Pkwy,144,Larrabee St & Webster Ave,Customer,NaN,NaN
4,23479392,2019-07-01 00:02:13,2019-07-01 00:22:26,6014,"1,213.0",168,Michigan Ave & 14th St,62,McCormick Place,Customer,NaN,NaN


In [ ]:
#check how many has commas in q3
df_q3_clean['tripduration'].str.contains(',', na=False).sum()

np.int64(667557)

In [ ]:
#check how many has commas in q4
df_q4_clean['tripduration'].str.contains(',', na=False).sum()

np.int64(176266)

In [ ]:
#check how many has commas in q2
df_q2_clean['tripduration'].str.contains(',', na=False).sum()

np.int64(407348)

In [ ]:
# Data Cleaning Function for 2019 Data set

def clean_2019_data(df):
    
    # Create working copy
    df = df.copy()
    
    # Convert date columns
    df['start_time'] = pd.to_datetime(df['start_time'])
    df['end_time'] = pd.to_datetime(df['end_time'])
    
    # Clean tripduration

    ## Remove commas
    df['tripduration'] = (
        df['tripduration']
        .str.replace(',', '', regex=False)
    )
    
    ## convert tripduration column
    df['tripduration'] = pd.to_numeric(df['tripduration'])
    
    return df

In [ ]:
# Clean q1 data set
q1_2019_clean = clean_2019_data(df)

# Clean q2 data set
q2_2019_clean = clean_2019_data(df_q2_clean)

# Clean q3 data set
q3_2019_clean = clean_2019_data(df_q3_clean)

# Clean q4 data set
q4_2019_clean = clean_2019_data(df_q4_clean)

In [ ]:
# check that the data has been cleaned and the data types converted
q4_2019_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 704054 entries, 0 to 704053
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   trip_id            704054 non-null  int64         
 1   start_time         704054 non-null  datetime64[ns]
 2   end_time           704054 non-null  datetime64[ns]
 3   bikeid             704054 non-null  int64         
 4   tripduration       704054 non-null  float64       
 5   from_station_id    704054 non-null  int64         
 6   from_station_name  704054 non-null  object        
 7   to_station_id      704054 non-null  int64         
 8   to_station_name    704054 non-null  object        
 9   usertype           704054 non-null  object        
 10  gender             637463 non-null  object        
 11  birthyear          642373 non-null  float64       
dtypes: datetime64[ns](2), float64(2), int64(4), object(4)
memory usage: 64.5+ MB


In [ ]:
# final check just to be sure every column matches
q4_2019_clean.columns.tolist()

['trip_id',
 'start_time',
 'end_time',
 'bikeid',
 'tripduration',
 'from_station_id',
 'from_station_name',
 'to_station_id',
 'to_station_name',
 'usertype',
 'gender',
 'birthyear']

In [ ]:
#Finally combine all 2019 datasets into a single dataset
all_2019 = pd.concat(
    [
        q1_2019_clean,
        q2_2019_clean,
        q3_2019_clean,
        q4_2019_clean
    ],
    ignore_index=True
)

In [ ]:
# Check newly created dataset
all_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818004 entries, 0 to 3818003
Data columns (total 12 columns):
 #   Column             Dtype         
---  ------             -----         
 0   trip_id            int64         
 1   start_time         datetime64[ns]
 2   end_time           datetime64[ns]
 3   bikeid             int64         
 4   tripduration       float64       
 5   from_station_id    int64         
 6   from_station_name  object        
 7   to_station_id      int64         
 8   to_station_name    object        
 9   usertype           object        
 10  gender             object        
 11  birthyear          float64       
dtypes: datetime64[ns](2), float64(2), int64(4), object(4)
memory usage: 349.5+ MB


In [ ]:
all_2019.shape

(3818004, 12)

In [ ]:
df_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426887 entries, 0 to 426886
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             426887 non-null  object 
 1   rideable_type       426887 non-null  object 
 2   started_at          426887 non-null  object 
 3   ended_at            426887 non-null  object 
 4   start_station_name  426887 non-null  object 
 5   start_station_id    426887 non-null  int64  
 6   end_station_name    426886 non-null  object 
 7   end_station_id      426886 non-null  float64
 8   start_lat           426887 non-null  float64
 9   start_lng           426887 non-null  float64
 10  end_lat             426886 non-null  float64
 11  end_lng             426886 non-null  float64
 12  member_casual       426887 non-null  object 
dtypes: float64(5), int64(1), object(7)
memory usage: 42.3+ MB


In [ ]:
# create a copy of 2020 dataset
df_2020_clean = df_2020.copy()

In [ ]:
df_2020_clean['started_at'] = pd.to_datetime(df_2020_clean['started_at'])

df_2020_clean['ended_at'] = pd.to_datetime(df_2020_clean['ended_at'])

In [ ]:
df_2020_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426887 entries, 0 to 426886
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             426887 non-null  object        
 1   rideable_type       426887 non-null  object        
 2   started_at          426887 non-null  datetime64[ns]
 3   ended_at            426887 non-null  datetime64[ns]
 4   start_station_name  426887 non-null  object        
 5   start_station_id    426887 non-null  int64         
 6   end_station_name    426886 non-null  object        
 7   end_station_id      426886 non-null  float64       
 8   start_lat           426887 non-null  float64       
 9   start_lng           426887 non-null  float64       
 10  end_lat             426886 non-null  float64       
 11  end_lng             426886 non-null  float64       
 12  member_casual       426887 non-null  object        
dtypes: datetime64[ns](2), float64

In [ ]:
# create a tripduration column for the 2020 dateset
df_2020_clean['tripduration'] = (df_2020_clean['ended_at'] - df_2020_clean['started_at']).dt.total_seconds()

In [ ]:
# confirm that tripduration column was created and the valies were in seconds and the datatype is float
df_2020_clean['tripduration'].head()

0    451.0
1    223.0
2    171.0
3    529.0
4    332.0
Name: tripduration, dtype: float64

In [ ]:
# one more check to see all the columns together
df_2020_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426887 entries, 0 to 426886
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             426887 non-null  object        
 1   rideable_type       426887 non-null  object        
 2   started_at          426887 non-null  datetime64[ns]
 3   ended_at            426887 non-null  datetime64[ns]
 4   start_station_name  426887 non-null  object        
 5   start_station_id    426887 non-null  int64         
 6   end_station_name    426886 non-null  object        
 7   end_station_id      426886 non-null  float64       
 8   start_lat           426887 non-null  float64       
 9   start_lng           426887 non-null  float64       
 10  end_lat             426886 non-null  float64       
 11  end_lng             426886 non-null  float64       
 12  member_casual       426887 non-null  object        
 13  tripduration        426887 no

2020 and 2019 have similar fields but use different names e.g for 2020 we have 'ride_id' and for 2019 we have 'bike_id' se we need to 
standardize the 2020 dataset column names so that we would be able to merge it with 2019

In [ ]:
# rename 2020 dataset column names to match that of 2019
df_2020_clean = df_2020_clean.rename(
    columns={
        'ride_id': 'trip_id',
        'started_at': 'start_time',
        'ended_at': 'end_time',
        'member_casual': 'usertype'
    }
)

In [118]:
# confirm the renaming 
df_2020_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426887 entries, 0 to 426886
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   trip_id             426887 non-null  object        
 1   rideable_type       426887 non-null  object        
 2   start_time          426887 non-null  datetime64[ns]
 3   end_time            426887 non-null  datetime64[ns]
 4   start_station_name  426887 non-null  object        
 5   start_station_id    426887 non-null  int64         
 6   end_station_name    426886 non-null  object        
 7   end_station_id      426886 non-null  float64       
 8   start_lat           426887 non-null  float64       
 9   start_lng           426887 non-null  float64       
 10  end_lat             426886 non-null  float64       
 11  end_lng             426886 non-null  float64       
 12  usertype            426887 non-null  object        
 13  tripduration        426887 no

In [ ]:
# Check if there are multiple rideable_types to determine it's usefulness in this analysis
df_2020_clean['rideable_type'].value_counts()

rideable_type
docked_bike    426887
Name: count, dtype: int64

In [120]:
# Drop rideable_type because it contains only one unique value
df_2020_clean = df_2020_clean.drop(columns=['rideable_type'])

## Standardize naming for both datasets so we can join both

In [123]:
# create a copy of all 2019 data set
all_2019_final = all_2019.copy()

In [124]:
# Change the names of the columns in 2019 to 2020
all_2019_final = all_2019_final.rename(
    columns={
        'from_station_id': 'start_station_id',
        'from_station_name': 'start_station_name',
        'to_station_id': 'end_station_id',
        'to_station_name': 'end_station_name'
    }
)

In [125]:
# confirm changes
all_2019_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818004 entries, 0 to 3818003
Data columns (total 12 columns):
 #   Column              Dtype         
---  ------              -----         
 0   trip_id             int64         
 1   start_time          datetime64[ns]
 2   end_time            datetime64[ns]
 3   bikeid              int64         
 4   tripduration        float64       
 5   start_station_id    int64         
 6   start_station_name  object        
 7   end_station_id      int64         
 8   end_station_name    object        
 9   usertype            object        
 10  gender              object        
 11  birthyear           float64       
dtypes: datetime64[ns](2), float64(2), int64(4), object(4)
memory usage: 349.5+ MB


In [126]:
all_2019_final.columns.to_list()

['trip_id',
 'start_time',
 'end_time',
 'bikeid',
 'tripduration',
 'start_station_id',
 'start_station_name',
 'end_station_id',
 'end_station_name',
 'usertype',
 'gender',
 'birthyear']

In [127]:
df_2020_clean.columns.to_list()

['trip_id',
 'start_time',
 'end_time',
 'start_station_name',
 'start_station_id',
 'end_station_name',
 'end_station_id',
 'start_lat',
 'start_lng',
 'end_lat',
 'end_lng',
 'usertype',
 'tripduration']

In [128]:
common_columns = [
    'trip_id',
    'start_time',
    'end_time',
    'tripduration',
    'start_station_id',
    'start_station_name',
    'end_station_id',
    'end_station_name',
    'usertype'
]

all_2019_final = all_2019_final[common_columns]

In [129]:
df_2020_final = df_2020_clean[common_columns]

In [130]:
df_2020_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426887 entries, 0 to 426886
Data columns (total 9 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   trip_id             426887 non-null  object        
 1   start_time          426887 non-null  datetime64[ns]
 2   end_time            426887 non-null  datetime64[ns]
 3   tripduration        426887 non-null  float64       
 4   start_station_id    426887 non-null  int64         
 5   start_station_name  426887 non-null  object        
 6   end_station_id      426886 non-null  float64       
 7   end_station_name    426886 non-null  object        
 8   usertype            426887 non-null  object        
dtypes: datetime64[ns](2), float64(2), int64(1), object(4)
memory usage: 29.3+ MB


In [131]:
all_2019_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818004 entries, 0 to 3818003
Data columns (total 9 columns):
 #   Column              Dtype         
---  ------              -----         
 0   trip_id             int64         
 1   start_time          datetime64[ns]
 2   end_time            datetime64[ns]
 3   tripduration        float64       
 4   start_station_id    int64         
 5   start_station_name  object        
 6   end_station_id      int64         
 7   end_station_name    object        
 8   usertype            object        
dtypes: datetime64[ns](2), float64(1), int64(3), object(3)
memory usage: 262.2+ MB


In [132]:
# merge both data sets together
all_trips = pd.concat(
    [all_2019_final, df_2020_final],
    ignore_index=True
)

In [133]:
# Confirm merge
all_trips.shape

(4244891, 9)

In [134]:
all_trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4244891 entries, 0 to 4244890
Data columns (total 9 columns):
 #   Column              Dtype         
---  ------              -----         
 0   trip_id             object        
 1   start_time          datetime64[ns]
 2   end_time            datetime64[ns]
 3   tripduration        float64       
 4   start_station_id    int64         
 5   start_station_name  object        
 6   end_station_id      float64       
 7   end_station_name    object        
 8   usertype            object        
dtypes: datetime64[ns](2), float64(2), int64(1), object(4)
memory usage: 291.5+ MB


In [135]:
# Check for duplicates
all_trips['trip_id'].duplicated().sum()

np.int64(0)

In [136]:
# check for null values
all_trips.isnull().sum()

trip_id               0
start_time            0
end_time              0
tripduration          0
start_station_id      0
start_station_name    0
end_station_id        1
end_station_name      1
usertype              0
dtype: int64

In [137]:
# confirm where values are missing
all_trips[
    all_trips['end_station_name'].isnull()
]

,trip_id,start_time,end_time,tripduration,start_station_id,start_station_name,end_station_id,end_station_name,usertype
4232430,157EAA4C4A3C8D36,2020-03-16 11:23:36,2020-03-16 11:23:24,-12.0,675,HQ QR,NaN,NaN,casual


In [138]:
# since it's only a single row out of 4 million which makes it insignificant we drop it
all_trips = all_trips.dropna(
    subset=['end_station_id', 'end_station_name']
)

In [139]:
# confirm there are no more null values
all_trips.isnull().sum()

trip_id               0
start_time            0
end_time              0
tripduration          0
start_station_id      0
start_station_name    0
end_station_id        0
end_station_name      0
usertype              0
dtype: int64

In [140]:
all_trips.head()

,trip_id,start_time,end_time,tripduration,start_station_id,start_station_name,end_station_id,end_station_name,usertype
0,21742443,2019-01-01 00:04:37,2019-01-01 00:11:07,390.0,199,Wabash Ave & Grand Ave,84.0,Milwaukee Ave & Grand Ave,Subscriber
1,21742444,2019-01-01 00:08:13,2019-01-01 00:15:34,441.0,44,State St & Randolph St,624.0,Dearborn St & Van Buren St (*),Subscriber
2,21742445,2019-01-01 00:13:23,2019-01-01 00:27:12,829.0,15,Racine Ave & 18th St,644.0,Western Ave & Fillmore St (*),Subscriber
3,21742446,2019-01-01 00:13:45,2019-01-01 00:43:28,1783.0,123,California Ave & Milwaukee Ave,176.0,Clark St & Elm St,Subscriber
4,21742447,2019-01-01 00:14:52,2019-01-01 00:20:56,364.0,173,Mies van der Rohe Way & Chicago Ave,35.0,Streeter Dr & Grand Ave,Subscriber


In [ ]:
# since tripduration is already in seconds let us check to ensure it all matches up before creating  ride_length_seconds
(
    (all_trips['end_time'] - all_trips['start_time'])
    .dt.total_seconds()
    != all_trips['tripduration']
).sum()

np.int64(1112296)

In [142]:
# Check why there is a mismatch
mismatch = all_trips[
    (all_trips['end_time'] - all_trips['start_time']).dt.total_seconds()
    != all_trips['tripduration']
]

mismatch.head()

,trip_id,start_time,end_time,tripduration,start_station_id,start_station_name,end_station_id,end_station_name,usertype
19374,21764881,2019-01-05 12:02:09,2019-01-19 18:01:51,1231180.0,59,Wabash Ave & Roosevelt Rd,24.0,Fairbanks Ct & Grand Ave,Subscriber
70751,21823975,2019-01-16 08:58:05,2019-02-05 19:12:07,1764840.0,142,McClurg Ct & Erie St,47.0,State St & Kinzie St,Customer
86090,21842257,2019-01-20 10:47:56,2019-02-19 11:02:35,2592880.0,627,LaSalle Dr & Huron St (*),360.0,DIVVY Map Frame B/C Station,Subscriber
88324,21845123,2019-01-21 23:07:19,2019-03-28 08:30:17,5646180.0,390,Wentworth Ave & 63rd St,120.0,Wentworth Ave & Cermak Rd (Temp),Subscriber
114411,21878086,2019-02-04 18:41:48,2019-03-12 15:36:20,3095670.0,110,Dearborn St & Erie St,283.0,LaSalle St & Jackson Blvd,Subscriber


In [143]:
# compare directly
mismatch[
    ['start_time', 'end_time', 'tripduration']
].head()

,start_time,end_time,tripduration
19374,2019-01-05 12:02:09,2019-01-19 18:01:51,1231180.0
70751,2019-01-16 08:58:05,2019-02-05 19:12:07,1764840.0
86090,2019-01-20 10:47:56,2019-02-19 11:02:35,2592880.0
88324,2019-01-21 23:07:19,2019-03-28 08:30:17,5646180.0
114411,2019-02-04 18:41:48,2019-03-12 15:36:20,3095670.0


In [144]:
mismatch['calculated_duration'] = (
    mismatch['end_time'] - mismatch['start_time']
).dt.total_seconds()

mismatch[
    ['tripduration', 'calculated_duration']
].head(10)

C:\Users\HEY\AppData\Local\Temp\ipykernel_9976\4222483903.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mismatch['calculated_duration'] = (


,tripduration,calculated_duration
19374,1231180.0,1231182.0
70751,1764840.0,1764842.0
86090,2592880.0,2592879.0
88324,5646180.0,5649778.0
114411,3095670.0,3099272.0
128296,4859470.0,4863068.0
132676,3926850.0,3930450.0
134262,1368310.0,1368311.0
148897,10628400.0,10632022.0
205358,819486.0,823086.0


In [145]:
# end station became a float instead of int because of the missing value convert back to int
all_trips['end_station_id'] = (
    all_trips['end_station_id']
    .astype('int64')
)

In [146]:
duration_diff = (
    (all_trips['end_time'] - all_trips['start_time']).dt.total_seconds()
    - all_trips['tripduration']
)

duration_diff.describe()

count    4.244890e+06
mean     2.230027e-01
std      1.523920e+01
min     -3.600000e+03
25%      0.000000e+00
50%      0.000000e+00
75%      1.000000e+00
max      3.622000e+03
dtype: float64

In [147]:
large_diff = all_trips[
    duration_diff.abs() >= 3600
]

large_diff.shape

(46, 9)

In [148]:
large_diff[
    ['start_time', 'end_time', 'tripduration']
].head(10)

,start_time,end_time,tripduration
114411,2019-02-04 18:41:48,2019-03-12 15:36:20,3095670.0
132676,2019-02-08 21:59:45,2019-03-26 09:47:15,3926850.0
148897,2019-02-14 14:44:13,2019-06-17 16:04:35,10628400.0
205358,2019-03-02 12:22:03,2019-03-12 01:00:09,819486.0
230157,2019-03-09 12:28:35,2019-03-10 16:35:42,97627.0
230755,2019-03-09 17:09:49,2019-03-10 03:34:08,33859.0
230836,2019-03-09 19:08:53,2019-03-10 14:52:53,67440.0
230915,2019-03-09 20:30:51,2019-03-10 10:30:12,46761.0
231084,2019-03-10 01:40:07,2019-03-10 03:10:28,1821.0
231089,2019-03-10 01:43:19,2019-03-10 03:19:37,2178.0


In [150]:
all_trips['year'] = all_trips['start_time'].dt.year

all_trips['month'] = all_trips['start_time'].dt.month

all_trips['month_name'] = all_trips['start_time'].dt.month_name()

all_trips['day'] = all_trips['start_time'].dt.day

all_trips['day_of_week'] = all_trips['start_time'].dt.dayofweek

all_trips['day_name'] = all_trips['start_time'].dt.day_name()

all_trips['hour'] = all_trips['start_time'].dt.hour

all_trips['is_weekend'] = all_trips['day_of_week'].isin([5, 6])

all_trips['tripduration_minutes'] = all_trips['tripduration'] / 60

In [151]:
all_trips.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4244890 entries, 0 to 4244890
Data columns (total 18 columns):
 #   Column                Dtype         
---  ------                -----         
 0   trip_id               object        
 1   start_time            datetime64[ns]
 2   end_time              datetime64[ns]
 3   tripduration          float64       
 4   start_station_id      int64         
 5   start_station_name    object        
 6   end_station_id        int64         
 7   end_station_name      object        
 8   usertype              object        
 9   year                  int32         
 10  month                 int32         
 11  month_name            object        
 12  day                   int32         
 13  day_of_week           int32         
 14  day_name              object        
 15  hour                  int32         
 16  is_weekend            bool          
 17  tripduration_minutes  float64       
dtypes: bool(1), datetime64[ns](2), float64(2), int3

In [152]:
all_trips['tripduration_hours'] = all_trips['tripduration'] / 3600

In [153]:
all_trips[['tripduration',
           'tripduration_minutes',
           'tripduration_hours']].head()

,tripduration,tripduration_minutes,tripduration_hours
0,390.0,6.500000,0.108333
1,441.0,7.350000,0.122500
2,829.0,13.816667,0.230278
3,1783.0,29.716667,0.495278
4,364.0,6.066667,0.101111


In [154]:
all_trips['usertype'].value_counts()

usertype
Subscriber    2937367
Customer       880637
member         378407
casual          48479
Name: count, dtype: int64

In [155]:
round(
    all_trips['usertype'].value_counts(normalize=True) * 100,
    2
)

usertype
Subscriber    69.20
Customer      20.75
member         8.91
casual         1.14
Name: proportion, dtype: float64

In [156]:
all_trips['tripduration_minutes'].describe()

count    4.244890e+06
mean     2.396341e+01
std      5.110926e+02
min     -9.200000e+00
25%      6.683333e+00
50%      1.150000e+01
75%      2.081667e+01
max      1.771400e+05
Name: tripduration_minutes, dtype: float64

In [157]:
all_trips['year'].value_counts()

year
2019    3818004
2020     426886
Name: count, dtype: int64

In [158]:
all_trips['month_name'].value_counts().sort_index()

month_name
April        265310
August       590184
December     155092
February     235771
January      247156
July         557315
June         475395
March        309028
May          367458
November     177176
October      371786
September    493219
Name: count, dtype: int64

In [159]:
(all_trips['tripduration'] < 0).sum()

np.int64(116)

In [160]:
all_trips = all_trips[
    all_trips['tripduration'] >= 0
]

In [161]:
(all_trips['tripduration'] < 0).sum()

np.int64(0)

In [162]:
all_trips['tripduration_minutes'].describe()

count    4.244774e+06
mean     2.396407e+01
std      5.110996e+02
min      0.000000e+00
25%      6.683333e+00
50%      1.150000e+01
75%      2.081667e+01
max      1.771400e+05
Name: tripduration_minutes, dtype: float64

In [163]:
all_trips.shape

(4244774, 19)

In [164]:
all_trips.columns.tolist()

['trip_id',
 'start_time',
 'end_time',
 'tripduration',
 'start_station_id',
 'start_station_name',
 'end_station_id',
 'end_station_name',
 'usertype',
 'year',
 'month',
 'month_name',
 'day',
 'day_of_week',
 'day_name',
 'hour',
 'is_weekend',
 'tripduration_minutes',
 'tripduration_hours']

In [165]:
all_trips['usertype'] = all_trips['usertype'].replace({
    'Subscriber': 'member',
    'Customer': 'casual'
})

In [166]:
all_trips['usertype'].value_counts()

usertype
member    3315774
casual     929000
Name: count, dtype: int64

In [167]:
# export cleaned dataset 
all_trips.to_csv(
    "../2. Data/cleaned/all_trips_cleaned.csv",
    index=False
)

In [168]:
import os

os.path.exists("../2. Data/cleaned/all_trips_cleaned.csv")

True

In [169]:
import pandas as pd

test_df = pd.read_csv(
    "../2. Data/cleaned/all_trips_cleaned.csv",
    nrows=5
)

test_df.head()

,trip_id,start_time,end_time,tripduration,start_station_id,start_station_name,end_station_id,end_station_name,usertype,year,month,month_name,day,day_of_week,day_name,hour,is_weekend,tripduration_minutes,tripduration_hours
0,21742443,2019-01-01 00:04:37,2019-01-01 00:11:07,390.0,199,Wabash Ave & Grand Ave,84,Milwaukee Ave & Grand Ave,member,2019,1,January,1,1,Tuesday,0,False,6.500000,0.108333
1,21742444,2019-01-01 00:08:13,2019-01-01 00:15:34,441.0,44,State St & Randolph St,624,Dearborn St & Van Buren St (*),member,2019,1,January,1,1,Tuesday,0,False,7.350000,0.122500
2,21742445,2019-01-01 00:13:23,2019-01-01 00:27:12,829.0,15,Racine Ave & 18th St,644,Western Ave & Fillmore St (*),member,2019,1,January,1,1,Tuesday,0,False,13.816667,0.230278
3,21742446,2019-01-01 00:13:45,2019-01-01 00:43:28,1783.0,123,California Ave & Milwaukee Ave,176,Clark St & Elm St,member,2019,1,January,1,1,Tuesday,0,False,29.716667,0.495278
4,21742447,2019-01-01 00:14:52,2019-01-01 00:20:56,364.0,173,Mies van der Rohe Way & Chicago Ave,35,Streeter Dr & Grand Ave,member,2019,1,January,1,1,Tuesday,0,False,6.066667,0.101111


In [170]:
all_trips.dtypes

trip_id                         object
start_time              datetime64[ns]
end_time                datetime64[ns]
tripduration                   float64
start_station_id                 int64
start_station_name              object
end_station_id                   int64
end_station_name                object
usertype                        object
year                             int32
month                            int32
month_name                      object
day                              int32
day_of_week                      int32
day_name                        object
hour                             int32
is_weekend                        bool
tripduration_minutes           float64
tripduration_hours             float64
dtype: object

In [171]:
all_trips.shape

(4244774, 19)

## Final Dataset Summary

After cleaning, feature engineering, and data quality validation:

- Rows: 4,244,774
- Columns: 19
- Missing values: 0
- Negative trip durations removed: 116
- Feature engineering completed:
  - year
  - month
  - month_name
  - day
  - day_of_week
  - day_name
  - hour
  - is_weekend
  - tripduration_minutes
  - tripduration_hours

Dataset exported for SQL analysis in PostgreSQL.